File: packages/less/src/less/parser/parser-input.test.js
import { describe, it, expect, beforeEach } from 'vitest';
import parserInputFactory from './parser-input';

describe('parser-input', () => {
    let parserInput;

    beforeEach(() => {
        parserInput = parserInputFactory();
    });

    describe('initialization', () => {
        it('should initialize with default values', () => {
            expect(parserInput.autoCommentAbsorb).toBe(true);
            expect(parserInput.commentStore).toEqual([]);
            expect(parserInput.finished).toBe(false);
        });
    });

    describe('start method', () => {
        it('should initialize with input string', () => {
            parserInput.start('test string');
            expect(parserInput.getInput()).toBe('test string');
            expect(parserInput.i).toBe(0);
        });

        it('should skip leading whitespace on start', () => {
            parserInput.start('  test string');
            expect(parserInput.i).toBe(2);
        });

        it('should handle chunked input', () => {
            parserInput.start('test string', true);
            expect(parserInput.getInput()).toBe('test string');
        });
    });

    describe('$re method', () => {
        beforeEach(() => {
            parserInput.start('test123');
        });

        it('should match regex pattern', () => {
            const result = parserInput.$re(/test/);
            expect(result).toBe('test');
            expect(parserInput.i).toBe(4);
        });

        it('should return null for no match', () => {
            const result = parserInput.$re(/xyz/);
            expect(result).toBeNull();
            expect(parserInput.i).toBe(0);
        });
    });

    describe('$char method', () => {
        beforeEach(() => {
            parserInput.start('test');
        });

        it('should match single character', () => {
            const result = parserInput.$char('t');
            expect(result).toBe('t');
            expect(parserInput.i).toBe(1);
        });

        it('should return null for no match', () => {
            const result = parserInput.$char('x');
            expect(result).toBeNull();
            expect(parserInput.i).toBe(0);
        });
    });

    describe('$str method', () => {
        beforeEach(() => {
            parserInput.start('test string');
        });

        it('should match exact string', () => {
            const result = parserInput.$str('test');
            expect(result).toBe('test');
            expect(parserInput.i).toBe(5);
        });

        it('should return null for no match', () => {
            const result = parserInput.$str('xyz');
            expect(result).toBeNull();
            expect(parserInput.i).toBe(0);
        });
    });

    describe('$quoted method', () => {
        it('should parse single quoted string', () => {
            parserInput.start("'test'");
            const result = parserInput.$quoted();
            expect(result).toBe("'test'");
            expect(parserInput.i).toBe(6);
        });

        it('should parse double quoted string', () => {
            parserInput.start('"test"');
            const result = parserInput.$quoted();
            expect(result).toBe('"test"');
            expect(parserInput.i).toBe(6);
        });

        it('should handle escaped quotes', () => {
            parserInput.start('"test\\"test"');
            const result = parserInput.$quoted();
            expect(result).toBe('"test\\"test"');
            expect(parserInput.i).toBe(12);
        });

        it('should return null for unclosed quotes', () => {
            parserInput.start('"test');
            const result = parserInput.$quoted();
            expect(result).toBeNull();
            expect(parserInput.i).toBe(0);
        });
    });

    describe('$parseUntil method', () => {
        it('should parse until matching character', () => {
            parserInput.start('test;end');
            const result = parserInput.$parseUntil(';');
            expect(result).toEqual(['test']);
            expect(parserInput.i).toBe(4);
        });

        it('should return null if token not found', () => {
            parserInput.start('test { block }');
            const result = parserInput.$parseUntil(';');
            expect(result).toBeNull();
        });
    });

    describe('peek methods', () => {
        beforeEach(() => {
            parserInput.start('test');
        });

        it('should peek at next character without consuming', () => {
            expect(parserInput.peek('t')).toBe(true);
            expect(parserInput.i).toBe(0);
        });

        it('should peek at current character', () => {
            expect(parserInput.currentChar()).toBe('t');
        });

        it('should peek at previous character', () => {
            parserInput.i = 1;
            expect(parserInput.prevChar()).toBe('t');
        });
    });

    describe('save and restore', () => {
        beforeEach(() => {
            parserInput.start('test string');
        });

        it('should save and restore parser state', () => {
            parserInput.save();
            parserInput.i = 5;
            parserInput.restore();
            expect(parserInput.i).toBe(0);
        });

        it('should track furthest position with error message', () => {
            parserInput.save();
            parserInput.i = 5;
            parserInput.restore('test error');
            expect(parserInput.i).toBe(0);
            const endState = parserInput.end();
            expect(endState.furthest).toBe(5);
            expect(endState.furthestPossibleErrorMessage).toBe('test error');
        });

        it('should forget saved state', () => {
            parserInput.save();
            parserInput.i = 5;
            parserInput.forget();
            expect(() => parserInput.restore()).toThrow();
        });
    });

    describe('end method', () => {
        it('should return correct end state when finished', () => {
            parserInput.start('test');
            parserInput.i = 4;
            const result = parserInput.end();
            expect(result.isFinished).toBe(true);
            expect(result.furthest).toBe(4);
        });

        it('should return correct end state when not finished', () => {
            parserInput.start('test string');
            parserInput.i = 4;
            const result = parserInput.end();
            expect(result.isFinished).toBe(false);
            expect(result.furthest).toBe(4);
        });

        it('should return furthest position with error message if error occurred', () => {
            parserInput.start('test');
            parserInput.save();
            parserInput.i = 2;
            parserInput.restore('test error');
            const result = parserInput.end();
            expect(result.isFinished).toBe(false);
            expect(result.furthest).toBe(2);
            expect(result.furthestPossibleErrorMessage).toBe('test error');
        });
    });

    describe('isWhitespace method', () => {
        it('should identify whitespace characters using offset', () => {
            const testString = 'test \t\n\r';
            parserInput.start(testString);
            expect(parserInput.i).toBe(0);
            expect(parserInput.isWhitespace(4)).toBe(true);
            expect(parserInput.isWhitespace(5)).toBe(true);
            expect(parserInput.isWhitespace(6)).toBe(true);
            expect(parserInput.isWhitespace(7)).toBe(true);
        });

        it('should identify non-whitespace characters', () => {
            const testString = 'test';
            parserInput.start(testString);
            expect(parserInput.i).toBe(0);
            expect(parserInput.isWhitespace(0)).toBe(false);
            expect(parserInput.isWhitespace(3)).toBe(false);
        });
    });

    describe('peekNotNumeric method', () => {
        it('should identify non-numeric characters', () => {
            parserInput.start('test');
            expect(parserInput.peekNotNumeric()).toBe(true);
        });

        it('should identify numeric characters', () => {
            parserInput.start('123');
            expect(parserInput.peekNotNumeric()).toBe(false);
        });

        it('should identify numeric characters with sign', () => {
            parserInput.start('+123');
            expect(parserInput.peekNotNumeric()).toBe(false);
            parserInput.start('-123');
            expect(parserInput.peekNotNumeric()).toBe(false);
        });

        it('should identify special characters as non-numeric', () => {
            parserInput.start('/123');
            expect(parserInput.peekNotNumeric()).toBe(true);
            parserInput.start(',123');
            expect(parserInput.peekNotNumeric()).toBe(true);
            parserInput.start('.123');
            expect(parserInput.peekNotNumeric()).toBe(false);
        });
    });

    describe('comment handling', () => {
        it('should handle line comments with autoCommentAbsorb enabled', () => {
            parserInput.start('test // comment\nend');
            const result = parserInput.$re(/test/);
            expect(result).toBe('test');
            expect(parserInput.i).toBe(16);
            expect(parserInput.commentStore).toHaveLength(1);
            expect(parserInput.commentStore[0].isLineComment).toBe(true);
            expect(parserInput.commentStore[0].text).toBe('// comment');
        });

        it('should handle block comments with autoCommentAbsorb enabled', () => {
            parserInput.start('test /* comment */ end');
            const result = parserInput.$re(/test/);
            expect(result).toBe('test');
            expect(parserInput.i).toBe(19);
            expect(parserInput.commentStore).toHaveLength(1);
            expect(parserInput.commentStore[0].isLineComment).toBe(false);
            expect(parserInput.commentStore[0].text).toBe('/* comment */');
        });

        it('should handle comments with autoCommentAbsorb disabled', () => {
            parserInput.autoCommentAbsorb = false;
            parserInput.start('test // comment\nend');
            const result = parserInput.$re(/test/);
            expect(result).toBe('test');
            expect(parserInput.i).toBe(5);
            expect(parserInput.commentStore).toHaveLength(0);
        });
    });

    describe('chunking', () => {
        it('should handle input with multiple chunks', () => {
            parserInput.start('chunk1;chunk2;chunk3', true);
            expect(parserInput.getInput()).toBe('chunk1;chunk2;chunk3');
        });

        it('should handle special characters at chunk boundaries', () => {
            parserInput.start('{chunk1};{chunk2}', true);
            expect(parserInput.getInput()).toBe('{chunk1};{chunk2}');
        });
    });

    describe('error handling', () => {
        it('should handle unmatched quotes', () => {
            parserInput.start('"unmatched quote');
            const result = parserInput.$quoted();
            expect(result).toBeNull();
        });

        it('should handle unmatched brackets', () => {
            parserInput.start('{unmatched bracket');
            const result = parserInput.$parseUntil('}');
            expect(result).toBeNull();
        });
    });

    describe('edge cases', () => {
        it('should handle empty input string', () => {
            parserInput.start('');
            expect(parserInput.getInput()).toBe('');
            expect(parserInput.i).toBe(0);
        });

        it('should handle input with only whitespace', () => {
            parserInput.start(' \t\n\r');
            expect(parserInput.i).toBe(4);
        });

        it('should handle input with only comments', () => {
            parserInput.start('/* comment */ // comment');
            const result = parserInput.$re(/.*/);
            expect(result).toBe('');
            expect(parserInput.i).toBe(25);
            expect(parserInput.commentStore).toHaveLength(2);
            expect(parserInput.commentStore[0].isLineComment).toBe(false);
            expect(parserInput.commentStore[0].text).toBe('/* comment */');
            expect(parserInput.commentStore[1].isLineComment).toBe(true);
            expect(parserInput.commentStore[1].text).toBe('// comment');
        });
    });

    describe('state management', () => {
        it('should handle multiple save/restore operations', () => {
            parserInput.start('test string');
            parserInput.save();
            parserInput.i = 5;
            parserInput.save();
            parserInput.i = 10;
            parserInput.restore();
            expect(parserInput.i).toBe(5);
            parserInput.restore();
            expect(parserInput.i).toBe(0);
        });

        it('should handle save/restore with different error messages', () => {
            parserInput.start('test string');
            parserInput.save();
            parserInput.i = 5;
            parserInput.restore('error1');
            parserInput.save();
            parserInput.i = 10;
            parserInput.restore('error2');
            const endState = parserInput.end();
            expect(endState.furthest).toBe(10);
            expect(endState.furthestPossibleErrorMessage).toBe('error2');
        });
    });

    describe('$parseUntil method', () => {
        it('should handle nested blocks', () => {
            parserInput.start('test { nested { block } } end');
            const result = parserInput.$parseUntil(';');
            expect(result).toBeNull();
        });

        it('should handle quoted strings inside blocks', () => {
            parserInput.start('test { "quoted" } end');
            const result = parserInput.$parseUntil(';');
            expect(result).toBeNull();
        });

        it('should handle comments inside blocks', () => {
            parserInput.start('test { /* comment */ } end');
            const result = parserInput.$parseUntil(';');
            expect(result).toBeNull();
        });
    });

    describe('whitespace handling', () => {
        it('should handle different whitespace combinations', () => {
            parserInput.start(' \t\n\r test');
            expect(parserInput.i).toBe(5);
        });

        it('should handle whitespace at boundaries', () => {
            parserInput.start(' test ');
            expect(parserInput.i).toBe(1);
        });

        it('should preserve whitespace inside quoted strings', () => {
            parserInput.start('" test "');
            const result = parserInput.$quoted();
            expect(result).toBe('" test "');
        });
    });

    describe('character code handling', () => {
        it('should handle ASCII boundary characters', () => {
            parserInput.start('\u0000\u007F');
            expect(parserInput.getInput()).toBe('\u0000\u007F');
        });

        it('should handle special characters', () => {
            parserInput.start('\\"\'`');
            expect(parserInput.getInput()).toBe('\\"\'`');
        });

        it('should handle non-breaking space', () => {
            parserInput.start('\u00A0');
            expect(parserInput.isWhitespace(0)).toBe(false);
        });
    });
});

File: packages/less/src/less/go_parser/parser_input_test.go
package go_parser

import (
	"regexp"
	"testing"
)

func TestParserInput(t *testing.T) {
	t.Run("initialization", func(t *testing.T) {
		parserInput := NewParserInput()
		if !parserInput.autoCommentAbsorb {
			t.Error("autoCommentAbsorb should be true by default")
		}
		if len(parserInput.commentStore) != 0 {
			t.Error("commentStore should be empty by default")
		}
		if parserInput.finished {
			t.Error("finished should be false by default")
		}
	})

	t.Run("start method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test string", false, nil)
		if parserInput.GetInput() != "test string" {
			t.Error("input should be set correctly")
		}
		if parserInput.i != 0 {
			t.Error("i should be 0")
		}

		parserInput = NewParserInput()
		parserInput.Start("  test string", false, nil)
		if parserInput.i != 2 {
			t.Error("should skip leading whitespace")
		}

		parserInput = NewParserInput()
		parserInput.Start("test string", true, nil)
		if parserInput.GetInput() != "test string" {
			t.Error("should handle chunked input")
		}
	})

	t.Run("$re method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test123", false, nil)

		result := parserInput.Re(regexp.MustCompile("test"))
		if result != "test" {
			t.Error("should match regex pattern")
		}
		if parserInput.i != 4 {
			t.Error("should advance position")
		}

		parserInput = NewParserInput()
		parserInput.Start("test123", false, nil)
		result = parserInput.Re(regexp.MustCompile("xyz"))
		if result != nil {
			t.Error("should return nil for no match")
		}
		if parserInput.i != 0 {
			t.Error("should not advance position")
		}
	})

	t.Run("$char method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test", false, nil)

		result := parserInput.Char('t')
		if resByte, ok := result.(byte); !ok || resByte != 't' {
			t.Error("should match single character")
		}
		if parserInput.i != 1 {
			t.Error("should advance position")
		}

		parserInput = NewParserInput()
		parserInput.Start("test", false, nil)
		result = parserInput.Char('x')
		if result != nil {
			t.Error("should return nil for no match")
		}
		if parserInput.i != 0 {
			t.Error("should not advance position")
		}
	})

	t.Run("$str method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test string", false, nil)

		result := parserInput.Str("test")
		if result != "test" {
			t.Error("should match exact string")
		}
		if parserInput.i != 5 {
			t.Error("should advance position")
		}

		parserInput = NewParserInput()
		parserInput.Start("test string", false, nil)
		result = parserInput.Str("xyz")
		if result != nil {
			t.Error("should return nil for no match")
		}
		if parserInput.i != 0 {
			t.Error("should not advance position")
		}
	})

	t.Run("$quoted method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("'test'", false, nil)
		result := parserInput.Quoted(-1)
		if result != "'test'" {
			t.Error("should parse single quoted string")
		}
		if parserInput.i != 6 {
			t.Error("should advance position")
		}

		parserInput = NewParserInput()
		parserInput.Start("\"test\"", false, nil)
		result = parserInput.Quoted(-1)
		if result != "\"test\"" {
			t.Error("should parse double quoted string")
		}
		if parserInput.i != 6 {
			t.Error("should advance position")
		}

		parserInput = NewParserInput()
		parserInput.Start("\"test\\\"test\"", false, nil)
		result = parserInput.Quoted(-1)
		if result != "\"test\\\"test\"" {
			t.Error("should handle escaped quotes")
		}
		if parserInput.i != 12 {
			t.Error("should advance position")
		}

		parserInput = NewParserInput()
		parserInput.Start("\"test", false, nil)
		result = parserInput.Quoted(-1)
		if result != nil {
			t.Error("should return nil for unclosed quotes")
		}
		if parserInput.i != 0 {
			t.Error("should not advance position")
		}
	})

	t.Run("$parseUntil method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test;end", false, nil)
		result := parserInput.ParseUntil(";")
		if result == nil {
			t.Error("should parse until matching character")
		}
		if parserInput.i != 4 {
			t.Error("should advance position")
		}

		parserInput = NewParserInput()
		parserInput.Start("test { block }", false, nil)
		result = parserInput.ParseUntil(";")
		if result != nil {
			t.Error("should return nil if token not found")
		}
	})

	t.Run("peek methods", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test", false, nil)

		if !parserInput.Peek("t") {
			t.Error("should peek at next character without consuming")
		}
		if parserInput.i != 0 {
			t.Error("should not advance position")
		}

		if parserInput.CurrentChar() != 't' {
			t.Error("should peek at current character")
		}

		parserInput.i = 1
		if parserInput.PrevChar() != 't' {
			t.Error("should peek at previous character")
		}
	})

	t.Run("save and restore", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test string", false, nil)

		parserInput.Save()
		parserInput.i = 5
		parserInput.Restore("")
		if parserInput.i != 0 {
			t.Error("should restore parser state")
		}

		parserInput = NewParserInput()
		parserInput.Start("test string", false, nil)
		parserInput.Save()
		parserInput.i = 5
		parserInput.Restore("test error")
		if parserInput.i != 0 {
			t.Error("should restore parser state")
		}
		endState := parserInput.End()
		if endState.Furthest != 5 {
			t.Error("should track furthest position")
		}
		if endState.FurthestPossibleErrorMessage != "test error" {
			t.Error("should track error message")
		}

		parserInput = NewParserInput()
		parserInput.Start("test string", false, nil)
		parserInput.Save()
		parserInput.i = 5
		parserInput.Forget()
		if len(parserInput.saveStack) != 0 {
			t.Error("should forget saved state")
		}
	})

	t.Run("end method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test", false, nil)
		parserInput.i = 4
		result := parserInput.End()
		if !result.IsFinished {
			t.Error("should return correct end state when finished")
		}
		if result.Furthest != 4 {
			t.Error("should return correct furthest position")
		}

		parserInput = NewParserInput()
		parserInput.Start("test string", false, nil)
		parserInput.i = 4
		result = parserInput.End()
		if result.IsFinished {
			t.Error("should return correct end state when not finished")
		}
		if result.Furthest != 4 {
			t.Error("should return correct furthest position")
		}

		parserInput = NewParserInput()
		parserInput.Start("test", false, nil)
		parserInput.Save()
		parserInput.i = 2
		parserInput.Restore("test error")
		result = parserInput.End()
		if result.IsFinished {
			t.Error("should return correct end state when error occurred")
		}
		if result.Furthest != 2 {
			t.Error("should return correct furthest position")
		}
		if result.FurthestPossibleErrorMessage != "test error" {
			t.Error("should return correct error message")
		}
	})

	t.Run("isWhitespace method", func(t *testing.T) {
		parserInput := NewParserInput()
		testString := "test \t\n\r"
		parserInput.Start(testString, false, nil)
		if parserInput.i != 0 {
			t.Error("i should be 0")
		}
		if !parserInput.IsWhitespace(4) {
			t.Error("should identify space as whitespace")
		}
		if !parserInput.IsWhitespace(5) {
			t.Error("should identify tab as whitespace")
		}
		if !parserInput.IsWhitespace(6) {
			t.Error("should identify newline as whitespace")
		}
		if !parserInput.IsWhitespace(7) {
			t.Error("should identify carriage return as whitespace")
		}

		parserInput = NewParserInput()
		parserInput.Start("test", false, nil)
		if parserInput.i != 0 {
			t.Error("i should be 0")
		}
		if parserInput.IsWhitespace(0) {
			t.Error("should identify non-whitespace characters")
		}
		if parserInput.IsWhitespace(3) {
			t.Error("should identify non-whitespace characters")
		}
	})

	t.Run("peekNotNumeric method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test", false, nil)
		if !parserInput.PeekNotNumeric() {
			t.Error("should identify non-numeric characters")
		}

		parserInput = NewParserInput()
		parserInput.Start("123", false, nil)
		if parserInput.PeekNotNumeric() {
			t.Error("should identify numeric characters")
		}

		parserInput = NewParserInput()
		parserInput.Start("+123", false, nil)
		if parserInput.PeekNotNumeric() {
			t.Error("should identify numeric characters with sign")
		}
		parserInput = NewParserInput()
		parserInput.Start("-123", false, nil)
		if parserInput.PeekNotNumeric() {
			t.Error("should identify numeric characters with sign")
		}

		parserInput = NewParserInput()
		parserInput.Start("/123", false, nil)
		if !parserInput.PeekNotNumeric() {
			t.Error("should identify special characters as non-numeric")
		}
		parserInput = NewParserInput()
		parserInput.Start(",123", false, nil)
		if !parserInput.PeekNotNumeric() {
			t.Error("should identify special characters as non-numeric")
		}
		parserInput = NewParserInput()
		parserInput.Start(".123", false, nil)
		if parserInput.PeekNotNumeric() {
			t.Error("should identify decimal point as numeric")
		}
	})

	t.Run("comment handling", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test // comment\nend", false, nil)
		result := parserInput.Re(regexp.MustCompile("test"))
		if result != "test" {
			t.Error("should handle line comments")
		}
		if parserInput.i != 16 {
			t.Error("should advance position past comment")
		}
		if len(parserInput.commentStore) != 1 {
			t.Error("should store comment")
		}
		if !parserInput.commentStore[0].isLineComment {
			t.Error("should identify line comment")
		}
		if parserInput.commentStore[0].text != "// comment" {
			t.Error("should store comment text")
		}

		parserInput = NewParserInput()
		parserInput.Start("test /* comment */ end", false, nil)
		result = parserInput.Re(regexp.MustCompile("test"))
		if result != "test" {
			t.Error("should handle block comments")
		}
		if parserInput.i != 19 {
			t.Error("should advance position past comment")
		}
		if len(parserInput.commentStore) != 1 {
			t.Error("should store comment")
		}
		if parserInput.commentStore[0].isLineComment {
			t.Error("should identify block comment")
		}
		if parserInput.commentStore[0].text != "/* comment */" {
			t.Error("should store comment text")
		}

		parserInput = NewParserInput()
		parserInput.autoCommentAbsorb = false
		parserInput.Start("test // comment\nend", false, nil)
		result = parserInput.Re(regexp.MustCompile("test"))
		if result != "test" {
			t.Error("should handle comments with autoCommentAbsorb disabled")
		}
		if parserInput.i != 5 {
			t.Error("should not advance position past comment")
		}
		if len(parserInput.commentStore) != 0 {
			t.Error("should not store comment")
		}
	})

	t.Run("chunking", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("chunk1;chunk2;chunk3", true, nil)
		if parserInput.GetInput() != "chunk1;chunk2;chunk3" {
			t.Error("should handle input with multiple chunks")
		}

		parserInput = NewParserInput()
		parserInput.Start("{chunk1};{chunk2}", true, nil)
		if parserInput.GetInput() != "{chunk1};{chunk2}" {
			t.Error("should handle special characters at chunk boundaries")
		}
	})

	t.Run("error handling", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("\"unmatched quote", false, nil)
		result := parserInput.Quoted(-1)
		if result != nil {
			t.Error("should handle unmatched quotes")
		}

		parserInput = NewParserInput()
		parserInput.Start("{unmatched bracket", false, nil)
		result = parserInput.ParseUntil("}")
		if result != nil {
			t.Error("should handle unmatched brackets")
		}
	})

	t.Run("edge cases", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("", false, nil)
		if parserInput.GetInput() != "" {
			t.Error("should handle empty input string")
		}
		if parserInput.i != 0 {
			t.Error("should not advance position")
		}

		parserInput = NewParserInput()
		parserInput.Start(" \t\n\r", false, nil)
		if parserInput.i != 4 {
			t.Error("should handle input with only whitespace")
		}

		parserInput = NewParserInput()
		parserInput.Start("/* comment */ // comment", false, nil)
		result := parserInput.Re(regexp.MustCompile(".*"))
		if resStr, ok := result.(string); !ok || resStr != "" {
			t.Error("should handle input with only comments")
		}
		t.Logf("Current parser state before final check: i=%d", parserInput.i)
		if parserInput.i != 25 {
			t.Errorf("should advance position past comments, expected 25, got %d", parserInput.i)
		}
		if len(parserInput.commentStore) != 2 {
			t.Error("should store comments")
		}
		if parserInput.commentStore[0].text != "/* comment */" || parserInput.commentStore[0].isLineComment {
			t.Error("should store block comment correctly")
		}
		if parserInput.commentStore[1].text != "// comment" || !parserInput.commentStore[1].isLineComment {
			t.Error("should store line comment correctly")
		}
	})

	t.Run("state management", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test string", false, nil)
		parserInput.Save()
		parserInput.i = 5
		parserInput.Save()
		parserInput.i = 10
		parserInput.Restore("")
		if parserInput.i != 5 {
			t.Error("should handle multiple save/restore operations")
		}
		parserInput.Restore("")
		if parserInput.i != 0 {
			t.Error("should handle multiple save/restore operations")
		}

		parserInput = NewParserInput()
		parserInput.Start("test string", false, nil)
		parserInput.Save()
		parserInput.i = 5
		parserInput.Restore("error1")
		parserInput.Save()
		parserInput.i = 10
		parserInput.Restore("error2")
		endState := parserInput.End()
		if endState.Furthest != 10 {
			t.Error("should track furthest position")
		}
		if endState.FurthestPossibleErrorMessage != "error2" {
			t.Error("should track latest error message")
		}
	})

	t.Run("$parseUntil method", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("test { nested { block } } end", false, nil)
		result := parserInput.ParseUntil(";")
		if result != nil {
			t.Error("should handle nested blocks")
		}

		parserInput = NewParserInput()
		parserInput.Start("test { \"quoted\" } end", false, nil)
		result = parserInput.ParseUntil(";")
		if result != nil {
			t.Error("should handle quoted strings inside blocks")
		}

		parserInput = NewParserInput()
		parserInput.Start("test { /* comment */ } end", false, nil)
		result = parserInput.ParseUntil(";")
		if result != nil {
			t.Error("should handle comments inside blocks")
		}
	})

	t.Run("whitespace handling", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start(" \t\n\r test", false, nil)
		if parserInput.i != 5 {
			t.Error("should handle different whitespace combinations")
		}

		parserInput = NewParserInput()
		parserInput.Start(" test ", false, nil)
		if parserInput.i != 1 {
			t.Error("should handle whitespace at boundaries")
		}

		parserInput = NewParserInput()
		parserInput.Start("\" test \"", false, nil)
		result := parserInput.Quoted(-1)
		if result != "\" test \"" {
			t.Error("should preserve whitespace inside quoted strings")
		}
	})

	t.Run("character code handling", func(t *testing.T) {
		parserInput := NewParserInput()
		parserInput.Start("\u0000\u007F", false, nil)
		if parserInput.GetInput() != "\u0000\u007F" {
			t.Error("should handle ASCII boundary characters")
		}

		parserInput = NewParserInput()
		parserInput.Start("\\\"'`", false, nil)
		if parserInput.GetInput() != "\\\"'`" {
			t.Error("should handle special characters")
		}

		parserInput = NewParserInput()
		parserInput.Start("\u00A0", false, nil)
		if parserInput.IsWhitespace(0) {
			t.Error("should handle non-breaking space")
		}
	})
} 